### ISSUE TITLE
Incorrect error code when Gruntfile missing

### ISSUE THREAD

**Issue by paulhayes** [gruntjs/grunt#1459]
http://gruntjs.com/api/exit-codes says missing grunt file should return exit code 2. Possibly only happening in osX, haven't tested elsewhere.

I ran the following and got exit code 99:
```
$ grunt --no-color --gruntfile ["web/app/themes/roots/Gruntfile.js"] build
grunt-cli: The grunt command line interface. (v0.1.13)

Fatal error: Unable to find local grunt.

If you're seeing this message, either a Gruntfile wasn't found or grunt
hasn't been installed locally to your project. For more information about
installing and configuring grunt, please see the Getting Started guide:

http://gruntjs.com/getting-started
$ echo $?
99
```

**Comment by jonaldomo** [comment]
This error comes from the grunt-cli, here is the code that returns 99.  https://github.com/gruntjs/grunt-cli/blob/64e1b531a1a35d21d628e32b6a6b5521081b1184/bin/grunt#L40

The second part of the error message is probably your problem "grunt
hasn't been installed locally to your project".  Not that it can't find the Gruntfile.js.

**Comment by JKAussieSkater** [comment]
> **@jonaldomo said:**
> the error message is probably your problem "grunt hasn't been installed locally to your project". _**Not that it can't find the Gruntfile.js**_.

I agree. @shama, I don't think this is a bug.

@paulhayes, I'm on Mac OSX 10.11.3 El Capitan, and I was unable to reproduce your error, except when `grunt` itself isn't installed correctly. I strongly believe that this is the case with your project as well (`grunt` not installed properly).

Also, the syntax you've provided is incorrect, as there should not be square brackets: `grunt --gruntfile "web/app/themes/roots/Gruntfile.js"`

----

I'll post up my attempt to reproduce the unexpected results in the post below:

**Comment by JKAussieSkater** [comment]
# Minimum Configuration

Here's my project directory:

    GruntProjectTest/
    ├── node_modules/
    │   └── grunt/
    │       └── ...
    └── Subfolder/
        └── Gruntfile.js


Here is the contents of my `Gruntfile.js`

```js
module.exports = function (grunt) {
  'use strict';
  grunt.initConfig({});
  grunt.registerTask('default', ['Hello-World']);
  grunt.registerTask("Hello-World", function () {
    console.log('Hello, World!');
  });
  grunt.registerTask("Goodbye-World", function () {
    console.log('Goodbye, World!');
  });
};
```

# Test Time

### Test 1: grunt
As the `Gruntfile.js` is not stored at the project's root directory, `grunt` should fail with error code 2:

```
$ cd "/Users/JKAussieSkater/Documents/GruntProjectTest"
$ grunt
A valid Gruntfile could not be found. Please see the getting started guide for
more information on how to configure grunt: http://gruntjs.com/getting-started
Fatal error: Unable to find Gruntfile.
$ echo $?
2
```

**Test Passed:** The expected outcome occurred.

### Test 2: grunt --gruntfile 'Gruntfile.js'
This test is identical to the previous one, except we are manually specifying what's normally the default location of the `Gruntfile.js`.

As the `Gruntfile.js` is not stored at the project's root directory, `grunt` should fail with error code 2:

```
$ cd "/Users/JKAussieSkater/Documents/GruntProjectTest"
$ grunt --gruntfile 'Gruntfile.js'
A valid Gruntfile could not be found. Please see the getting started guide for
more information on how to configure grunt: http://gruntjs.com/getting-started
Fatal error: Unable to find Gruntfile.
$ echo $?
2
```

**Test Passed:** The expected outcome occurred.

### Test 3: grunt --gruntfile 'Subfolder/Gruntfile.js'
As we will now specify the correct location of the `Gruntfile.js`, Grunt should execute the default task, and log "Hello, World!"

```
$ cd "/Users/JKAussieSkater/Documents/GruntProjectTest"
$ grunt --gruntfile 'Subfolder/Gruntfile.js'
Running "Hello-World" task
Hello, World!

Done, without errors.
$ echo $?
0
```

**Test Passed:** The expected outcome occurred.

### Test 4: grunt --gruntfile 'Subfolder/Gruntfile.js' Goodbye-World
Now we will try instructing Grunt to run a specific task only. This should log "Goodbye, World!".

```
$ cd "/Users/JKAussieSkater/Documents/GruntProjectTest"
$ grunt --gruntfile 'Subfolder/Gruntfile.js' Goodbye-World
Running "Goodbye-World" task
Goodbye, World!

Done, without errors.
$ echo $?
0
```

**Test Passed:** The expected outcome occurred.

### Test 5: grunt --gruntfile ['Subfolder/Gruntfile.js'] Goodbye-World
We will now try the previous test once more, but with incorrect square-bracket syntax in the `--gruntfile` option. Because the syntax is incorrect, it will try to find a file path that includes square brackets, which should lead to a non-existing file. This should, in turn, generate an error code 2.

```
$ cd "/Users/JKAussieSkater/Documents/GruntProjectTest"
$ grunt --gruntfile ['Subfolder/Gruntfile.js'] Goodbye-World

Reading "Gruntfile.js]" Gruntfile...ERROR
Fatal error: Unable to find "/Users/JKAussieSkater/Documents/GruntProjectTest/[Subfolder/Gruntfile.js]" Gruntfile.
$ echo $?
2
```

**Test Passed:** The expected outcome occurred.

----

# Summary so far
Provided that Grunt is installed correctly, we are getting the expected error codes for when `Gruntfile.js` isn't found.

Let's continue some more tests of when Grunt is NOT found... However, I'm using the **very latest** commit to `grunt-cli`, which has a [more accurate error message](https://github.com/gruntjs/grunt-cli/commit/9ea84be2b094f61b856eb44e9fea1cba3c242537). Keep this in mind.

### Test 6: grunt --gruntfile ['Subfolder/Gruntfile.js'] Goodbye-World
Nearly identical to test 5, however we will first rename the `node_modules/grunt` directory, so that `grunt-cli` will be **unable to find local grunt**. This should result in error code 99:

```
$ cd "/Users/JKAussieSkater/Documents/GruntProjectTest"
$ mv "node_modules/grunt" "node_modules/grunt_renamed"
$ grunt --gruntfile ['Subfolder/Gruntfile.js'] Goodbye-World
grunt-cli: The grunt command line interface (v1.0.0-rc1)

Fatal error: Unable to find local grunt.

If you're seeing this message, grunt hasn't been installed locally to
your project. For more information about installing and configuring grunt,
please see the Getting Started guide:

http://gruntjs.com/getting-started
$ echo $?
99
```

**Test Passed:** The expected outcome occurred.

### Test 7: grunt --gruntfile 'Subfolder/Gruntfile.js' Goodbye-World
Even if we use the correct syntax for the `--gruntfile` option, this shouldn't change the fact that Grunt wasn't installed correctly in the first place. This should result in error code 99:

```
$ cd "/Users/JKAussieSkater/Documents/GruntProjectTest"
$ mv "node_modules/grunt" "node_modules/grunt_renamed"
$ grunt --gruntfile 'Subfolder/Gruntfile.js' Goodbye-World
grunt-cli: The grunt command line interface (v1.0.0-rc1)

Fatal error: Unable to find local grunt.

If you're seeing this message, grunt hasn't been installed locally to
your project. For more information about installing and configuring grunt,
please see the Getting Started guide:

http://gruntjs.com/getting-started
$ echo $?
99
```

**Test Passed:** The expected outcome occurred.

### Test 8: grunt
Just to ease the minds of the skeptics, let's move the `Gruntfile.js` to the default location of the project's root directory, and attempt this again with Grunt not installed in the correct location. This should result in error code 99:

```
$ cd "/Users/JKAussieSkater/Documents/GruntProjectTest"
$ mv "node_modules/grunt" "node_modules/grunt_renamed"
$ mv "Subfolder/Gruntfile.js" "Gruntfile.js"
$ grunt
grunt-cli: The grunt command line interface (v1.0.0-rc1)

Fatal error: Unable to find local grunt.

If you're seeing this message, grunt hasn't been installed locally to
your project. For more information about installing and configuring grunt,
please see the Getting Started guide:

http://gruntjs.com/getting-started
$ echo $?
99
```

**Test Passed:** The expected outcome occurred.

----

At least on my machine, the expected results occur, which make me believe the [misleading `grunt-cli` error message](https://github.com/gruntjs/grunt-cli/issues/94) has confused developers into thinking that Grunt itself is returning the wrong error code.

If you still believe this is a grunt bug, could you please post up your "minimum configuration", as I have, and the test you did to achieve the incorrect error code.

**Comment by shama** [comment]
![applause](https://cloud.githubusercontent.com/assets/99604/13099826/58c21c9e-d4eb-11e5-962b-14024636a7ea.gif)

Nice @JKAussieSkater! :)

**Comment by JKAussieSkater** [comment]
Thanks! Bit long though, so to get to the point:

**Abridged version:** See headings "Summary so far", "Test 7" and "Conclusion"
**tl;dr** All tests were successful, there's no bug.

**Comment by vladikoff** [comment]
:eyes:

**Comment by paulhayes** [comment]
Holy moly.

Yup, that's pretty conclusive. I tip my hat to you **JKAussieSkater**
